### ISSUE TITLE
xgboost dart dask fails while gbtree does not:  AttributeError: '_thread._local' object has no attribute 'execution_state'

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#6607]
```
2021-01-15 05:54:54,522 C: 68% D:836.0GB M:224.8GB NODE:REMOTE1     7243   INFO   | [XGBOOSTDARTDASK] model fit (963689, 208) start.
2021-01-15 05:54:55,605 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   INFO   | [XGBOOSTDARTDASK] model fit (963689, 208) end. Took 1.07165 seconds.
2021-01-15 05:54:55,608 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   WARNING| fit failure for XGBoostDartDaskModel failed, skipping.
2021-01-15 05:54:55,609 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   WARNING|   Stack_file: [./tmp/nidhi/h2oai_experiment_bc0a9fea-56f2-11eb-b26b-0ea86ce99368/67a7ef1e-2347-40d9-8fa6-66613aef156f.stack]
2021-01-15 05:54:55,609 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | WARNING: fit failure for XGBoostDartDaskModel failed, skipping.  Error: [Traceback (most recent call last):
2021-01-15 05:54:55,610 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/worker.py", line 3072, in get_worker
2021-01-15 05:54:55,610 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     return thread_state.execution_state["worker"]
2021-01-15 05:54:55,611 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | AttributeError: '_thread._local' object has no attribute 'execution_state'
2021-01-15 05:54:55,611 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,611 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | During handling of the above exception, another exception occurred:
2021-01-15 05:54:55,612 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,612 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | Traceback (most recent call last):
2021-01-15 05:54:55,613 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/worker.py", line 3075, in get_worker
2021-01-15 05:54:55,613 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     return first(w for w in Worker._instances if w.status == "running")
2021-01-15 05:54:55,613 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "cytoolz/itertoolz.pyx", line 550, in cytoolz.itertoolz.first
2021-01-15 05:54:55,614 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "cytoolz/itertoolz.pyx", line 557, in cytoolz.itertoolz.first
2021-01-15 05:54:55,614 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | StopIteration
2021-01-15 05:54:55,615 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,615 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | During handling of the above exception, another exception occurred:
2021-01-15 05:54:55,615 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,616 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | Traceback (most recent call last):
2021-01-15 05:54:55,616 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "h2oaicore/dask.py", line 97, in h2oaicore.dask.get_dask_client_manager
2021-01-15 05:54:55,617 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/python/lib/python3.6/contextlib.py", line 81, in __enter__
2021-01-15 05:54:55,617 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     return next(self.gen)
2021-01-15 05:54:55,617 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/worker_client.py", line 41, in worker_client
2021-01-15 05:54:55,618 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     worker = get_worker()
2021-01-15 05:54:55,618 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/worker.py", line 3077, in get_worker
2021-01-15 05:54:55,619 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     raise ValueError("No workers found")
2021-01-15 05:54:55,619 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | ValueError: No workers found
2021-01-15 05:54:55,619 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,620 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | During handling of the above exception, another exception occurred:
2021-01-15 05:54:55,620 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | 
2021-01-15 05:54:55,620 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | Traceback (most recent call last):
1886, in h2oaicore.models.MainModel.dask_fit
2021-01-15 05:54:55,624 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "h2oaicore/dask.py", line 179, in h2oaicore.dask.get_client_manager
2021-01-15 05:54:55,624 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "h2oaicore/dask.py", line 102, in h2oaicore.dask.get_dask_client_manager
2021-01-15 05:54:55,624 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "h2oaicore/dask.py", line 85, in h2oaicore.dask.get_dask_client
2021-01-15 05:54:55,625 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/worker.py", line 3118, in get_client
2021-01-15 05:54:55,625 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     address = comm.resolve_address(address)
2021-01-15 05:54:55,626 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/comm/addressing.py", line 174, in resolve_address
2021-01-15 05:54:55,626 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     return unparse_address(scheme, backend.resolve_address(loc))
2021-01-15 05:54:55,626 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/comm/tcp.py", line 533, in resolve_address
2021-01-15 05:54:55,627 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     return unparse_host_port(ensure_ip(host), port)
2021-01-15 05:54:55,627 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "cytoolz/functoolz.pyx", line 476, in cytoolz.functoolz._memoize.__call__
2021-01-15 05:54:55,628 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/utils.py", line 711, in ensure_ip
2021-01-15 05:54:55,628 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     raise exc
2021-01-15 05:54:55,628 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/cuda-10.0/lib/python3.6/site-packages/distributed/utils.py", line 704, in ensure_ip
2021-01-15 05:54:55,629 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     hostname, 1234, fam, socket.SOCK_STREAM  # dummy port number
2021-01-15 05:54:55,629 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |   File "/opt/h2oai/dai/python/lib/python3.6/socket.py", line 745, in getaddrinfo
2021-01-15 05:54:55,630 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   |     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
2021-01-15 05:54:55,630 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | socket.gaierror: [Errno -2] Name or service not known
2021-01-15 05:54:55,630 C: 24% D:836.1GB M:225.0GB NODE:REMOTE1     7243   DATA   | ].
```

@trivialfis  Odd thing here is only happens when using cluster but just one node.  However, I'm running the client from a separate docker instance to the scheduler/workers.  There are 4 GPUs so 4 workers.  This is on ec2 system.  The processes look like:
```
root       7243  5.9  0.2 4427060 713376 ?      Sl   Jan14  25:32          \_ h2oai-server
root       7580  1.8  0.0 395592 204788 ?       Sl   Jan14   7:49              \_ dask-scheduler [tcp://10.10.0.20:8786]
root       7581  1.7  0.1 1335968 303512 ?      Sl   Jan14   7:38              \_ dask-scheduler [tcp://10.10.0.20:8790]
root       8058  2.4  0.2 5860772 623180 ?      Sl   Jan14  10:27              \_ h2oai_worker
root       8172  1.1  0.0 291864 60032 ?        Sl   Jan14   5:03                  \_ /opt/h2oai/dai/python/bin/python /opt/h2oai/dai/python/bin/dask-worker tcp://10.10.0.20:8786 --pid-file ./tmp/dai_dask_worker_n_jobs_2021-01-14_23_49_29.000901.pid --nthreads 1 --nprocs 1 --protocol tcp --resources n_jobs=1 --local-directory ./tmp/dask_worker_files
root       8190  0.0  0.0  26420 11080 ?        S    Jan14   0:00                  |   \_ /opt/h2oai/dai/python/bin/python -c from multiprocessing.semaphore_tracker import main;main(8)
root       8192  138  0.3 2079728 990540 ?      Sl   Jan14 630:04                  |   \_ dask-worker [tcp://10.10.0.20:46716]
root       8173  2.6  0.1 19749084 288752 ?     Sl   Jan14  12:13                  \_ /opt/h2oai/dai/python/bin/python /opt/h2oai/dai/python/bin/dask-cuda-worker tcp://10.10.0.20:8790 --pid-file ./tmp/dai_dask_cuda_worker_n_jobs_2021-01-14_23_49_29.000924.pid --nthreads 1 --resources n_gpus=1 n_jobs=1 --local-directory ./tmp/dask_cuda_worker_files
root       8287  0.0  0.0  26420 11120 ?        S    Jan14   0:00                  |   \_ /opt/h2oai/dai/python/bin/python -c from multiprocessing.semaphore_tracker import main;main(39)
root       8291  2.5  0.6 22073768 1699308 ?    Sl   Jan14  11:30                  |   \_ dask-worker [tcp://10.10.0.20:36126]
root       8293  2.5  0.6 22285456 1745476 ?    Sl   Jan14  11:31                  |   \_ dask-worker [tcp://10.10.0.20:35708]
root       8295  2.5  0.6 22183220 1691412 ?    Sl   Jan14  11:31                  |   \_ dask-worker [tcp://10.10.0.20:43614]
root       8298  2.5  0.6 22172068 1681012 ?    Sl   Jan14  11:29                  |   \_ dask-worker [tcp://10.10.0.20:42043]
```

Note I have 1 "CPU" scheduler and worker and 1 "GPU" with 4 workers launched by the dask-cuda-worker.  Here I'm running on GPU.

For some reason gbtree works perfectly.  However, dart fails every time with the above error.  Any ideas?

It's like dart somehow ruins the threading state, while gbtree does not.

**Comment by trivialfis** [comment]
> For some reason gbtree works perfectly. However, dart fails every time with the above error. Any ideas?

Probably empty dataset handling in dart, just guessing.

**Comment by trivialfis** [comment]
@pseudotensor Could you please try again after https://github.com/dmlc/xgboost/pull/6693 being merged?

**Comment by trivialfis** [comment]
Merged.  Now dart is running on CI.

**Comment by trivialfis** [comment]
We have some fixes and tests for dart in 1.4rc.  Feel free to reopen if the issue persists.
